Add `assert_regex`

rico-chet opened the pull request:
Add a function wrapping the commonly used `[[ value =~ pattern ]]`
construct, printing both operands in case of failure.

This addresses the need in #7 without affecting backward compatibility
of `assert_equal`.

martin-schulze-vireso commented:
I am unsure about potential code sharing with `assert_output --regex`.

martin-schulze-vireso commented:
Pleased add a test for the invalid regex branch.

martin-schulze-vireso commented on src/assert_regex.bash:
```diff
@@ -0,0 +1,41 @@
+
+	if [[ '' =~ ${pattern} ]] || (( ${?} == 2 )); then
+		echo "Invalid extended regular expression: \`${pattern}'" \
+		| batslib_decorate 'ERROR: assert_output' \
```
This should be assert_regex?

rico-chet commented:
I added the error case test and fixed a copy-paste issue.

I also found that the multi-line value case needs more test coverage and documentation. It's not obvious that the assert succeeds when _any_ line of the `value` matches the `pattern`.

rico-chet commented:
> > It's not obvious that the assert succeeds when _any_ line of the `value` matches the `pattern`.
> 
> I would have thought that it is default regex behavior to Match anywhere in the string If there are no begin/end markers ($^).

This is the case, but I don't know by heart how multi-line stuff is handled exactly, so I added a couple characterization tests.

rico-chet commented:
> I am unsure about potential code sharing with `assert_output --regex`. Then again its not even 20 lines...

`assert_output` only acts on the `$output` content, but not on e.g. `$stderr` or any other data like a call trace from a mock script (the scenario where I used this last time). This is the rationale for not re-using `assert_output` for this. Also, I don't like all the parameters as they clutter tests in the long run.

rico-chet commented:
should be good to go now

martin-schulze-vireso commented:
Should we also provide refute_regex in this PR?

rico-chet commented on README.md:
```diff
@@ -693,6 +693,62 @@ An error is displayed if the specified extended regular expression is invalid.
+If the value is longer than one line then it is displayed in *multi-line*
+format.
+
+Multi-line value matches a single-line pattern when one of its lines matches,
```
I get what you mean (`\n` in the value matches `.` in the pattern) and I was confused as I don't use multi-line matching often. The message needs an improvement, I agree. I will see how fast I can do it.

rico-chet replied:
I am planning to simply refer to the Bash docs: https://www.gnu.org/software/bash/manual/html_node/Conditional-Constructs.html#:~:text=[[…]], as it's rather complicated. The applied case sensitivity should also be made visible to users, IMHO. `BASH_REMATCH` visibility should be part of a test, too.

rico-chet commented:
> Should we also provide refute_regex in this PR?

I prefer a separate PR, to keep the batch size small and get things done faster.

martin-schulze-vireso commented:
Thanks for your contribution.
